File: server/app/schemas/book.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class Book(BaseModel):
    id: int
    title: str = Field(..., description="展示书名")
    author: Optional[str] = Field(None, description="原作者")
    cover_url: Optional[str] = Field(None, description="封面相对路径")
    file_name: str = Field(..., description="原始文件名")
    file_path: str = Field(..., description="图书文件相对路径")
    status: str = Field(..., description="图书状态")
    status_detail: Optional[str] = Field(None, description="状态说明")
    ai_friend_id: Optional[int] = Field(None, description="绑定好友 ID")
    reading_location: Optional[str] = Field(None, description="阅读定位值")
    create_time: datetime
    update_time: datetime
    deleted: bool
    file_size: Optional[int] = Field(None, description="文件大小（字节）")
    format_type: str = Field(..., description="格式类型")

    model_config = ConfigDict(from_attributes=True)

File: front/src/api/book.ts
import { withApiBase } from './base'

export interface Book {
  id: number
  title: string
  author?: string | null
  cover_url?: string | null
  file_name: string
  file_path: string
  status: string
  status_detail?: string | null
  ai_friend_id?: number | null
  reading_location?: string | null
  create_time: string
  update_time: string
  deleted: boolean
  file_size?: number | null
  format_type: string
}

async function parseError(response: Response, fallback: string): Promise<string> {
  try {
    const data = await response.json()
    if (typeof data?.detail === 'string' && data.detail.trim()) {
      return data.detail
    }
  } catch {
    // ignore json parse errors
  }
  return fallback
}

export async function getBooks(skip: number = 0, limit: number = 100): Promise<Book[]> {
  const params = new URLSearchParams({
    skip: skip.toString(),
    limit: limit.toString(),
  })
  const response = await fetch(withApiBase(`/api/books/?${params}`))
  if (!response.ok) {
    throw new Error(await parseError(response, '获取图书列表失败'))
  }
  return response.json()
}

export async function importBook(file: File): Promise<Book> {
  const formData = new FormData()
  formData.append('file', file)

  const response = await fetch(withApiBase('/api/books/import'), {
    method: 'POST',
    body: formData,
  })

  if (!response.ok) {
    throw new Error(await parseError(response, '导入图书失败'))
  }

  return response.json()
}
